sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1910,510 @@ def transform(self, X):
+        noise.
+
+    copy : boolean, optional, (default=True)
+        Set to False to perform inplace scaling and avoid a copy (if the input
+        is already a numpy array).
+
+    Attributes
+    ----------
+    quantiles_ : ndarray, shape (n_quantiles, n_features)
+        The values corresponding the quantiles of reference.
+
+    references_ : ndarray, shape(n_quantiles, )
+        Quantiles of references.
+
+    Examples
+    --------
+    >>> from sklearn.datasets import load_iris
+    >>> from sklearn.preprocessing import QuantileTransformer
+    >>> iris = load_iris()
+    >>> X_trans = QuantileTransformer(n_quantiles=20).fit_transform(iris.data)
+
+    See also
+    --------
+    quantile_transform : Equivalent function without the object oriented API.
+
+    StandardScaler : perform standardization that is faster, but less robust
+    to outliers.
+
+    RobustScaler : perform robust standardization that removes the influence
+    of outliers but does not put outliers and inliers on the same scale.
+
+    """
+
+    def __init__(self, n_quantiles=1000, output_distribution='uniform',
+                 ignore_implicit_zeros=False, subsample=int(1e5),
+                 smoothing_noise=None, random_state=None, copy=True):
+        self.n_quantiles = n_quantiles
+        self.output_distribution = output_distribution
+        self.ignore_implicit_zeros = ignore_implicit_zeros
+        self.subsample = subsample
+        self.smoothing_noise = smoothing_noise
+        self.random_state = random_state
+        self.copy = copy
+
+    def _compute_quantile_one_column(self, X_col, references, random_state):
+        """Private function to compute the quantiles for one features."""
+        if self.smoothing_noise is not None:
+            X_col += random_state.normal(0, self.smoothing_noise,
+                                         size=X_col.shape)
+
+        return np.percentile(X_col, references)
+
+    def _dense_fit(self, X, random_state):
+        """Compute percentiles for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+        """
+        if self.ignore_implicit_zeros:
+            warnings.warn("'ignore_implicit_zeros' takes effect only with"
+                          " sparse matrix. This parameter has no effect.")
+
+        n_samples, n_features = X.shape
+        # for compatibility issue with numpy<=1.8.X, references
+        # need to be a list scaled between 0 and 100
+        references = list(map(lambda x: x * 100, self.references_))
+        self.quantiles_ = []
+        for col in X.T:
+            if self.subsample < n_samples:
+                subsample_idx = choice(n_samples, size=self.subsample,
+                                       replace=False,
+                                       random_state=random_state)
+            else:
+                subsample_idx = range(n_samples)
```

Comment:
In this case, we should avoid a copying indexing operation, so rather than storing the `subsample_idx`, just do `col = col.take(...)` in the other case and make this case do nothing.